pkg/engine2/path_selection/candidate_weight_test.go

```diff
@@ -0,0 +1,159 @@
+package path_selection
+
+import (
+	"testing"
+
+	construct "github.com/klothoplatform/klotho/pkg/construct2"
+	"github.com/klothoplatform/klotho/pkg/construct2/graphtest"
+	"github.com/klothoplatform/klotho/pkg/engine2/enginetesting"
+	knowledgebase "github.com/klothoplatform/klotho/pkg/knowledge_base2"
+	"github.com/klothoplatform/klotho/pkg/knowledge_base2/kbtesting"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/mock"
+	"github.com/stretchr/testify/require"
+)
+
+func Test_determineCandidateWeight(t *testing.T) {
+	// NOTE(gg): this test is a little brittle, since the weights don't really have any meaning other than
+	// their relative values. They're made up to get the desired results in path selection.
+	tests := []struct {
+		name        string
+		graph       []any
+		resultGraph []any
+		src, target string
+		id          string
+		wantWeight  int
+		wantErr     bool
+	}{
+		{
+			name:       "no relation",
+			graph:      []any{"p:compute:a -> p:glue:b -> p:glue:c -> p:compute:d"},
+			src:        "p:compute:a",
+			target:     "p:compute:d",
+			id:         "p:compute:e",
+			wantWeight: 2,
+		},
+		{
+			name:        "no relation, in result graph",
+			graph:       []any{"p:compute:a -> p:glue:b -> p:glue:c -> p:compute:d"},
+			resultGraph: []any{"p:compute:e"},
+			src:         "p:compute:a",
+			target:      "p:compute:d",
+			id:          "p:compute:e",
+			wantWeight:  11,
+		},
+		{
+			name:       "downstream direct",
+			graph:      []any{"p:compute:a -> p:glue:b -> p:glue:c -> p:compute:d"},
+			src:        "p:compute:a",
+			target:     "p:compute:d",
+			id:         "p:glue:b",
+			wantWeight: 21,
+		},
+		{
+			name:       "downstream indirect",
+			graph:      []any{"p:compute:a -> p:glue:b -> p:glue:c -> p:glue:d -> p:compute:e"},
+			src:        "p:compute:a",
+			target:     "p:compute:e",
+			id:         "p:glue:c",
+			wantWeight: 15,
+		},
```

Comment:
should also add cases for in graph, but unrelated to downstream/upstream (checking that it uses the undirected graph to determine weight)

Reply:
👍 was planning on doing that (among other changes) for the +7% this week